internal/template-validator/validator/app.go

```diff
@@ -71,68 +79,122 @@ func (app *App) Run() {
 	informers.Start()
 	defer informers.Stop()
 
-	validating.NewWebhooks(informers).Register()
-
-	registerReadinessProbe()
+	if err := validatorMetrics.SetupMetrics(); err != nil {
+		logger.Log.Error(err, "Error setting up metrics")
+		panic(err)
+	}
 
-	// setup monitoring
-	err = validatorMetrics.SetupMetrics()
+	tlsInfo, err := app.setupTLS()
 	if err != nil {
-		logger.Log.Error(err, "Error setting up metrics")
 		panic(err)
 	}
+	if tlsInfo != nil {
+		defer tlsInfo.Clean()
+	}
 
-	http.Handle("/metrics", promhttp.Handler())
+	metricsServer := app.createMetricsServer()
+	webhookServer := app.createWebhookServer(informers)
+	if tlsInfo != nil {
+		metricsServer.TLSConfig = createTLSConfig(tlsInfo)
+		webhookServer.TLSConfig = createTLSConfig(tlsInfo)
+	}
 
-	if app.certsDir != "" {
-		logger.Log.Info("TLS certs directory", "directory", app.certsDir)
+	g := &errgroup.Group{}
+	g.Go(func() error { return startServer(metricsServer, metricsServerType) })
+	g.Go(func() error { return startServer(webhookServer, webhookServerType) })
 
-		tlsInfo := tlsinfo.TLSInfo{
-			CertsDirectory:      app.certsDir,
-			TLSOptionsDirectory: tlsOptionsDirectory,
-		}
+	if err := g.Wait(); err != nil {
+		panic(err)
+	}
 
-		if err := tlsInfo.Init(); err != nil {
-			logger.Log.Error(err, "Failed initializing TLSInfo")
-			panic(err)
-		}
-		defer tlsInfo.Clean()
+	shutdownServer(metricsServer, metricsServerType)
+	shutdownServer(webhookServer, webhookServerType)
```

Comment:
These calls are not needed. If this code is reached, then both of the serves have already cleanly shut down.

If there would be any error returned by the `startServer` the above panic would trigger.

Reply:
Updated